### ISSUE TITLE
AE error running single sample with external counts

### ISSUE THREAD

**Issue by muhligs** [gagneurlab/drop#461]
Hi, 
I am using drop v1.3.1
I am running a single local sample and external counts for 48 samples. 
I get the following error when I run the AE module:
[edit, not sure the config file uploaded correctly, but only AE is set and it looks like this:

```
projectTitle: "DROP: full run with own ext counts - one sample"
root: "/faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/Output"
htmlOutputPath: "/faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/Output/html"
indexWithFolderName: true # whether the root base name should be part of the index name

hpoFile: null  # if null, downloads it from webserver
sampleAnnotation: "/faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/Data/sample_annotation.tsv"

geneAnnotation:
    v43: "/faststorage/project/drop_rnaseq/sandbox/referencefiles/gencode.v43.basic.annotation.gtf"
genomeAssembly: hg38
genome: 
  somaseqref: "/faststorage/project/drop_rnaseq/sandbox/referencefiles/GCA_000001405.15_GRCh38_no_alt_analysis_set.GRC_exclusions_masked.fna"
    # You can define multiple reference genomes in yaml format, ncbi: path/to/ncbi, ucsc: path/to/ucsc
    # the keywords that define the path should be in the GENOME column of the sample annotation table

random_seed: true  # just for demo data, remove for analysis

exportCounts:
    # specify which gene annotations to include and which
    # groups to exclude when exporting counts
    geneAnnotations:
        - v43
#    excludeGroups:
#        - group2
#        - maegroup

aberrantExpression:
    run: true
    groups:
        - outrider
    fpkmCutoff: 1
    implementation: autoencoder
    padjCutoff: 0.05
    zScoreCutoff: 0
    genesToTest: null
    maxTestedDimensionProportion: 3
    dassie:
        tssWindow: 500
        pasWindow: 1000
    yieldSize: 2000000

```
]

```
Error in rule AberrantExpression_pipeline_Counting_Summary_R:
    jobid: 3
    input: /faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/Output/processed_results/aberrant_expression/v43/outrider/outrider/ods_unfitted.Rds, /faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/Output/processed_data/aberrant_expression/bam_stats/outrider.tsv, Scripts/AberrantExpression/pipeline/Counting/Summary.R
    output: /faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/Output/html/AberrantExpression/Counting/v43/Summary_outrider.html
    log: /faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/.drop/tmp/AE/v43/outrider/count_summary.Rds (check log file(s) for error details)
    cluster_jobid: Submitted batch job 17730787

Error executing rule AberrantExpression_pipeline_Counting_Summary_R on cluster (jobid: 3, external: Submitted batch job 17730787, jobscript: /faststorage/project/drop_rnaseq/sandbox/1b_sample_2023-05-08/.snakemake/tmp.id7_8klb/snakejob.AberrantExpression_pipeline_Counting_Summary_R.3.sh). For error details see the cluster log and the log files of the involved rule(s).

```

I have attached the sample_annotation and config files. 

Can you see what may be the issue?

Thanks, 
[config .txt](https://github.com/gagneurlab/d
[sample_annotation.txt](https://github.com/gagneurlab/drop/files/11421327/sample_annotation.txt)
rop/files/11421325/config.txt)

Morten

**Comment by muhligs** [comment]
A little more context for this error from the rule log file:
(Also of note, this happens both with our own external counts and with Gtex external counts. )

#######################################################################################
rule AberrantExpression_pipeline_Counting_Summary_R:
    input: /faststorage/project/drop_rnaseq/sandbox/1c_sample_2023-05-12/Output/processed_results/aberrant_expression/v29/outrider/outrider/ods_unfitted.Rds, /faststorage/project/drop_rnaseq/sandbox/1c_sample_2023-05-12/Output/processed_data/aberrant_expression/bam_stats/outrider.tsv, Scripts/AberrantExpression/pipeline/Counting/Summary.R
    output: /faststorage/project/drop_rnaseq/sandbox/1c_sample_2023-05-12/Output/html/AberrantExpression/Counting/v29/Summary_outrider.html
    log: /faststorage/project/drop_rnaseq/sandbox/1c_sample_2023-05-12/.drop/tmp/AE/v29/outrider/count_summary.Rds
    jobid: 0
    reason: Input files updated by another job: /faststorage/project/drop_rnaseq/sandbox/1c_sample_2023-05-12/Output/processed_data/aberrant_expression/bam_stats/outrider.tsv
    wildcards: annotation=v29, dataset=outrider
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/scratch/17937884

Loading required package: knitr
Loading required package: rmarkdown
Warning message:
package ‘rmarkdown’ was built under R version 4.2.3 
[1] TRUE


processing file: /scratch/17937884/RtmpWNAV0G/file7f07582aecba/Summary.spin.Rmd

  |                                                                                                  
  |                                                                                            |   0%
  |                                                                                                  
  |.......                                                                                     |   8%                  
  |                                                                                                  
  |..............                                                                              |  15% (unnamed-chunk-1)
  |                                                                                                  
  |.....................                                                                       |  23%                  
  |                                                                                                  
  |............................                                                                |  31% (unnamed-chunk-2)
  |                                                                                                  
  |...................................                                                         |  38%                  
  |                                                                                                  
  |..........................................                                                  |  46% (unnamed-chunk-3)Quitting from lines 113-164 (/scratch/17937884/RtmpWNAV0G/file7f07582aecba/Summary.spin.Rmd) 
Error in base::rowMeans(x, na.rm = na.rm, dims = dims, ...) : 
  'x' must be an array of at least two dimensions
Calls: render ... FUN -> data.table -> rowMeans -> rowMeans -> <Anonymous>
#######################################################################################

Any thoughts?

**Comment by muhligs** [comment]
It seems that the error  occurs because there is only one local sample. In the R script **Scripts/AberrantExpression/pipeline/Counting/Summary.R**
we get the error as illustrated below (around line 100 or so in the script)
The object `filter_mtx` contains the entries `local, all, passedFPKM, min 1 read, min 1 reads`
 The last four are matrices, but the local one is a vector, and going through the 
`data.table(gene_ID = rownames(mtx), median_counts = rowMeans(mtx), filter = filter_name)`
line below fails because rowMeans does not work on vectors. 

```
...
#' # Filtering
#' **local**: A pre-filtered summary of counts using only the local (from BAM) counts. Omitted if no external counts  
#' **all**: A pre-filtered summary of counts using only the merged local (from BAM) and external counts  
#' **passed FPKM**: Passes the user defined FPKM cutoff in at least 5% of genes  
#' **min 1 read**: minimum of 1 read expressed in 5% of genes  
#' **min 10 reads**: minimum of 10 reads expressed in 5% of genes  

quant <- .95

if(has_external){
    filter_mtx <- list(
      local = cnts_mtx_local,
      all = cnts_mtx,
      `passed FPKM` = cnts_mtx[rowData(ods)$passedFilter,],
      `min 1 read` = cnts_mtx[rowQuantiles(cnts_mtx, probs = quant) > 1, ],
      `min 10 reads` = cnts_mtx[rowQuantiles(cnts_mtx, probs = quant) > 10, ]
    ) # this runs
    
    
    filter_dt <- lapply(names(filter_mtx), function(filter_name) {
      mtx <- filter_mtx[[filter_name]]
      data.table(gene_ID = rownames(mtx), median_counts = rowMeans(mtx), filter = filter_name)
    }) %>% rbindlist # This does not run

#    [ my addition to troubleshoot
    mtx_1 <- filter_mtx[[names(filter_mtx)[1]]] # this is the 'local' 
    data.table(gene_ID = rownames(mtx_1), median_counts = rowMeans(mtx_1), filter = names(filter_mtx)[1])
    class(mtx_1) # integer (vector)
    mtx_2 <- filter_mtx[[names(filter_mtx)[2]]]
    data.table(gene_ID = rownames(mtx_2), median_counts = rowMeans(mtx_2), filter = names(filter_mtx)[2])
    class(mtx_2) # matrix
    # end of addition]

```

**Comment by mincej** [comment]
Hey, 
I had this same error here: https://github.com/gagneurlab/drop/issues/440#issuecomment-1498180316

The issue still persisted in 1.3.1, but was fixed in 1.3.3: https://github.com/gagneurlab/drop/issues/450#issuecomment-1518430281

Give that a try and see how it works

**Comment by muhligs** [comment]
Thanks @mincej , Looks like the exact same problem. will try out.

**Comment by vyepez88** [comment]
thanks @mincej for helping here! @muhligs please let me know if 1.3.3 solved your issue!

**Comment by muhligs** [comment]
The issue was solved with 1.3.3. Thanks for the help.
